opt_einsum/backends/shared.py

```diff
@@ -0,0 +1,116 @@
+import contextlib
+import numbers
+from collections import OrderedDict
+
+from ..parser import get_symbol
+from .dispatch import get_func
+
+_SHARING_STACK = []
+_CURRENT_BACKEND = []
+
+
+@contextlib.contextmanager
+def shared_intermediates(cache=None):
+    """Context in which contract intermediate results are shared.
+
+    Note that intermediate computations will not be garbage collected until
+    1. this context exits, and
+    2. the yielded cache is garbage collected (if it was captured).
+
+    Parameters
+    ----------
+    cache : dict
+        If specified, a user-stored dict in which intermediate results will
+        be stored. This can be used to interleave sharing contexts.
+
+    Returns
+    -------
+    cache : dict
+        A dictionary in which sharing results are stored. If ignored,
+        sharing results will be garbage collected when this context is
+        exited. This dict can be passed to another context to resume
+        sharing.
+    """
+    if cache is None:
+        cache = {}
+    _SHARING_STACK.append(cache)
+    yield cache
+    _SHARING_STACK.pop()
+
+
+@contextlib.contextmanager
+def handle_sharing(backend):
+    if _SHARING_STACK and not _CURRENT_BACKEND:
+        _CURRENT_BACKEND.append(backend)
+        yield 'opt_einsum.backends.shared'
+        _CURRENT_BACKEND.pop()
+    if backend == 'opt_einsum.backends.shared' and not _CURRENT_BACKEND:
+        raise ValueError('shared backend is available only via shared_intermediates')
+    yield backend
+
+
+def _alpha_canonicalize(equation):
+    """Alpha convert in an order-independent canonical way.
+    """
+    rename = OrderedDict()
+    for name in equation:
+        if name in ',->':
+            continue
+        if name not in rename:
+            rename[name] = get_symbol(len(rename))
+    return ''.join(rename.get(x, x) for x in equation)
+
+
+def transpose(a, axes):
+    backend = _CURRENT_BACKEND[0]
+    cache = _SHARING_STACK[-1]
+    cache['tensor', id(a)] = a
+
+    axes = tuple(axes)
+    key = 'transpose', backend, id(a), axes
+    if key in cache:
+        return cache[key]
+
+    result = get_func('transpose', backend)(a, axes)
+    cache[key] = result
+    return result
+
+
+def tensordot(x, y, axes=2):
+    backend = _CURRENT_BACKEND[0]
+    cache = _SHARING_STACK[-1]
```

Comment:
Can we normal order `x, y` here as well using `id` as below? With indices explicitly labeled I think this should be ok.

Reply:
I believe we can't normal-order `x,y` because their non-contracted dimensions are treated differently: `x`'s are on the left and `y`'s are on the right.

Reply:
One possibility would be to check for (x, y, axes) then (y, x, axes[::-1]) and if its the second just perform the transposition (which will be something like ``transpose(x, [2, 3, 4, 0, 1])``.

Reply:
Yeah, in this first PR I've implemented lookup modulo-commutativity but not modulo-transpose. I agree that we could do this modulo-transpose lookup here and also in the `einsum` part. But I'm still struggling to get tests to pass even in the current PR, so I'd be happy to limit this first PR to commutativity.

Reply:
+1 to getting something in, can we flag this and make an issue so that we do not forget about this point?

Reply:
Yeah absolutely - it's certainly not necessary for this PR. Good to keep track of these things so they can maybe be added them later however!